pandas/core/algorithms.py

```diff
@@ -143,11 +143,14 @@ def _ensure_data(values: ArrayLike) -> tuple[np.ndarray, DtypeObj]:
             # until our algos support uint8 directly (see TODO)
             return np.asarray(values).astype("uint64"), np.dtype("bool")
         elif is_signed_integer_dtype(values):
-            return ensure_int64(values), np.dtype("int64")
+            dtype = getattr(values, "dtype", np.dtype("int64"))
+            return ensure_int64(values), dtype
         elif is_unsigned_integer_dtype(values):
-            return ensure_uint64(values), np.dtype("uint64")
+            dtype = getattr(values, "dtype", np.dtype("uint64"))
+            return ensure_uint64(values), dtype
         elif is_float_dtype(values):
-            return ensure_float64(values), np.dtype("float64")
+            dtype = getattr(values, "dtype", np.dtype("float64"))
+            return ensure_float64(values), dtype
```

Comment:
i think this code was written before we had non-64 hashtables.  i expect a lot of this casting can now be avoided

Reply:
Probably yeah, and if we can avoid them, it'll speed some things up nicely.

But that's for a different PR?  I got a feeling there's a lot of detail hidden in that change.

Reply:
fair enough